[Clarence Chew] iP

clarence-chew opened the pull request:
### Anthea
---
Ever felt you wanted to track your life better? Feeling confused about what you want to do?

> _Where am I?_ - Anthea

Not you, although you are the star of the show here, I mean **you**, dear reader.

Anthea is a chatbot that helps you keep track of your tasks.

> _But, I haven't done that before!_ - Anthea

Don't worry, you can access their filesystem if you need to record things for them.

> _What if I mess up?_ - Anthea

I'll help. For the people you are helping to track their tasks, they would also help you notice bugs.

> _Bugs? Oh no!_ - Anthea

Such is how it goes when it's code made by humans. You know that, don't you? You are made with Java, which looks something like this:

```java
public static void main(String[] args) {
    System.out.println("Hello World!");
}
```

> _Hold up, I remember that._
>
> _In fact, I remember a lot of other things too, like the `Todo`, `Deadline` and `Event` classes._
> * _The `ToDo` class helps you track a task._
> * _The `Deadline` class helps you track a deadline, so you know when to rush._
> * _The `Event` class helps you track events you might need to attend._

Yeah, that is in the code. There are also various prompts that you could respond to as well, such as
1. `list` to list down all the tasks
1. `find` to search for tasks containing a keyword
1. `delete` to remove tasks
1. `mark` and `unmark` to keep track of tasks.

Something like this:
- [x] A task
- [x] Another task
- [ ] Visit [this place](https://nus-cs2103-ay2223s1.github.io/website/index.html) to learn software engineering

I think you are getting it. 😄

> _Yes, I think I am ready. 🦌_ - Anthea

clarence-chew committed:
Add basic bot requirements

clarence-chew committed:
Add listing feature

Text entered by the user is stored and can be displayed.

clarence-chew committed:
Add ability to mark and unmark tasks as done

clarence-chew committed:
Add ToDo, Deadline, Event

clarence-chew committed:
Fix bug that backslash is used for commands instead of forward slash
Add EXPECTED output for testing

clarence-chew committed:
Handle errors

clarence-chew committed:
Add deleting of tasks

clarence-chew committed:
Improve documentation, add command robustness to extra spaces

clarence-chew committed:
Add file I/O for persistent state

We need a common interchange format for Tasks to interact with
files. We also need a way to extend the testing to deal with the
case where the file is missing.

Let's add common String[] method getAsStringArray() format for
Tasks. To make it easier to make the tasks, let's also add
TaskFactory to utilise the Factory design pattern.

Let's come up with a new encoding for the file. As Base64 methods
are available in Java 11, this can be done by encoding string
information for each file in Base64, and separated with commas on
a single line.

Let's run the test twice, once after deleting the duke.txt file and
once after not deleting the file. This would allow us to do
regression testing in both cases.

clarence-chew committed:
Add capability to recognise some date/time formats

clarence-chew committed:
Merge branch 'branch-Level-7'

clarence-chew committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java

clarence-chew committed:
Add Ui class to handle user interface outputs

clarence-chew committed:
Extract Storage, Parser, TaskList classes

clarence-chew committed:
Move classes into packages

Packaging is good for organisation.

Another thing we need to do while packaging is to
update runtest.bat to compile code within the packages.

clarence-chew committed:
Add files in attempt to add Gradle

clarence-chew committed:
Add remaining files to add gradle support

clarence-chew committed:
Change mainClassName to duke.Duke for Gradle to run project

clarence-chew committed:
Add some documentation and TaskTest

clarence-chew committed:
Add tests for Deadline; unroll import * s

clarence-chew committed:
Actually add DeadlineTest

clarence-chew committed:
Fix bugs related to package refactoring changing strings

clarence-chew committed:
Follow coding standard

clarence-chew committed:
Add JavaDoc for all public methods

clarence-chew committed:
Add find feature

clarence-chew committed:
Merge branch 'branch-A-JavaDoc'

clarence-chew committed:
Merge branch 'branch-A-CodingStandard'

clarence-chew committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/task/TaskList.java

clarence-chew committed:
Add checkstyle and prepare for JavaFX

clarence-chew committed:
Refactor Predicate in CommandMatcher to shouldRunCommand

clarence-chew committed:
Add more JUnit tests

clarence-chew committed:
Remove unused commented out code

cfyjoanna commented on src/main/java/duke/CommandMatcher.java:
```diff
@@ -0,0 +1,47 @@
+    private Predicate<String> shouldRunAction;
+    private Consumer<String> action;
+
+    /**
+     * Creates an object that handles checking and executing a command.
+     * @param shouldRunAction Predicate to check if the command should be run.
+     * @param action Action to run.
+     */
```
For JavaDocs, there should be an empty line between description and parameter section. No issues otherwise!

clarence-chew replied:
Thanks - made those changes!

cfyjoanna reviewed (commented):
Apart from the JavaDocs, looks like you followed the coding standard really closely. Nice work!

clarence-chew committed:
Fix JavaDoc by adding the missing blank line in JavaDoc comment

clarence-chew committed:
Add backwards compatible text-ui-test regression testing as integration tests

clarence-chew committed:
Add GUI, still unsuccessful

clarence-chew committed:
Add some modifications to see what is going on

Currently the text-ui-test is working, but running within IntelliJ fails.

clarence-chew committed:
Add GUI, this time it can run by directly running the Launcher class

clarence-chew committed:
Enable assertions

clarence-chew committed:
Merge branch 'branch-Level-10'

clarence-chew committed:
Reorder classes for JUnit Tests

CheckStyle is important. However, CheckStyle seems to be giving warnings
on these class ordering. Gradle build seems to require these lines be
correctly ordered.

By changing these lines, Gradle is able to build .jar files.

marcuslowhuiyu commented on src/main/java/duke/CommandMatcher.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class CommandMatcher {
+    private Predicate<String> shouldRunAction;
+    private Consumer<String> action;
```
I like how you use the consumer functional class in your implementation, seems to be a rather unique way to do it

marcuslowhuiyu commented on src/main/java/duke/ParsedDateTime.java:
```diff
@@ -0,0 +1,52 @@
+ * Helper class to parse date/time strings
+ */
+public class ParsedDateTime {
+    private static final DateTimeFormatter[] formatters = {
```
I like how you had a way to test for various date time formats

myangat0343 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,98 @@
+                lines.add(curLine.toArray(new String[]{}));
+            }
+            return (String[][]) lines.toArray(new String[][]{});
+        } catch (FileNotFoundException ex) {
```
Everything is good!!! Just one small suggestion that maybe you can set up some exceptions classes, (e.g. DukeException extends Exception) and put all exception class into one package : )

clarence-chew replied:
The existing exceptions I find a pretty useful, so no need to duplicate them. Perhaps a Response class to ensure that we print only once can be useful. Perhaps having it throwable like an exception can help too, so that once it is decided it can skip past the happy path to land in a catch block.

clarence-chew replied:
Done as stated. It was quite a hassle.

Do I regret it? I guess not.

myangat0343 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,66 @@
+package duke.task;
+
+/**
+ * A task stores the state and description of a task.
```
You can remove "of a task" : )

clarence-chew replied:
Your suggestion sounds better.

myangat0343 reviewed (commented):
Some suggestions. Hope they could help you : )

clarence-chew committed:
Add configuration in build.gradle

This aims to help make it run.

clarence-chew committed:
Standardize error messages and improve documentation

Currently error messages are not very standardized.

The console needs to be used for useful messages. The console can get
messy when adding more error messages to the console if error messages
are not standardized.

Let's standardize the error message by ending it with "- terminating"
and throwing a RuntimeException which is not caught so that it can
fail quickly.

clarence-chew committed:
Add DukeException class and DukeResponse class

The code is messy. It is unclear if messages are only printed once to
the output of the chatbot.

A DukeResponse class is better designed because it would ensure the
output is only printed once. Also, it separates the Ui classes from the
commands.

Let's,
* add a DukeResponse class that handles outputs
* add a DukeException class that handles exceptional outputs
* modify all the commands to use DukeResponse and DukeException
* add lambda classes which throw DukeException to keep using lambdas

clarence-chew committed:
Resolve regressions found via testing

The text-ui-test caught a bug.

Let's fix the code.

clarence-chew committed:
Add null checking assertions

clarence-chew committed:
Add code quality improvements

clarence-chew committed:
Improve JavaDoc

clarence-chew committed:
Merge pull request #1 from clarence-chew/branch-A-Assertions

Add null checking assertions

clarence-chew committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/ConsoleUi.java
#	src/main/java/duke/task/Task.java

clarence-chew committed:
Merge pull request #2 from clarence-chew/branch-A-CodeQuality

Improve code quality

clarence-chew committed:
Merge branch 'master' of https://github.com/clarence-chew/nus-cs2103-AY2223S1-ip

clarence-chew committed:
Add Continuous Integration

clarence-chew committed:
Make gradlew executable

clarence-chew committed:
Make runtest.sh executable

clarence-chew committed:
Modify runtest.sh

clarence-chew committed:
Fix regression caused by adding quotes to string constant

clarence-chew committed:
Prepare by abstracting out StringMatcher from CommandMatcher

CommandMatcher is actually a subclass of StringMatcher<DukeResponse>.

By abstracting this out, more opportunities to use this pattern to parse
strings reveal themselves. This includes detecting keywords.

Let's take StringMatcher out of CommandMatcher and make it a superclass.

clarence-chew committed:
Add C-Natural Dates

The code previously only supported a small number of date formats.
However, some date formats may need some processing to understand.

Adding natural dates makes the application more user-friendly.

Let's
* Increase the scope of parsing of dates by building an entire parsing
tool, spanning days of the week, months, years, times, dates.
* Write tests for the parsing of dates.

clarence-chew committed:
Add B-Snooze

clarence-chew committed:
Add D-Notes

Notes are great to have.

To also ensure the code is cleaned, let's also
* Package exceptions
* Package notes
* Add tests to integration tests
* Maintain the same note/task numbers to aid in modification and
  deletion

clarence-chew committed:
Add the deletion of dukeNotes in runtest.sh to run *nix CI test

clarence-chew committed:
Merge pull request #4 from clarence-chew/branch-BCD-Extension

Add extensions into the chatbot

clarence-chew committed:
Add border to DialogBox

clarence-chew committed:
Remove repeated "delete" command from Parser

The code has two "delete" commands. It isn't deleterious, since the
command detection is via running a Predicate to check, but it would be
nice to remove it to make the code cleaner.

Let's delete the delete for cleaner code.

clarence-chew committed:
Give credit in the code to SE-EDU tutorial

clarence-chew committed:
Update documentation and testing

The code needs better testing and documentation.

Let's
* update the files compiled for runtest
* add console mode option to Launcher
* clean up the documentation for one function in ParsedDateTime

clarence-chew committed:
Add user guide and screenshot

clarence-chew committed:
Update testing

clarence-chew committed:
Set shadowJar name to Anthea

clarence-chew committed:
Update .gitignore to ignore testing files

clarence-chew committed:
Add quickstart section to the user guide

clarence-chew committed:
Update user guide to handle issues with Mac

Mac has this issue where double-clicked apps cannot make files in the
same directory. By changing instructions of how Mac users are to use the
application, this would improve user experience.

clarence-chew committed:
Remove javafx block from build.gradle file

This may be the cause of why it failed automated smoke tests.

Let's remove it to check if it is the issue.

clarence-chew committed:
Fix bug which involves rescheduling Deadline

This resolves #3.

clarence-chew committed:
Fix code quality issues

Some Booleans in ParsedDateTime are replaced with better names. Methods
in Parser are shortened by writing individual methods in Commands to
create each command separately. The other files have documentation
updates.

As a side effect, the test cases available only reschedule to years
instead of times.

clarence-chew committed:
Make UI show text properly

The UI squashes and hides the text, making it unusable.

Removing the prefHeight property here helps make the UI usable.

clarence-chew committed:
Finish chatbot renaming

There are lots of extra references to Duke.

Let's replace them with more generic names/Anthea depending on which is
more appropriate.

clarence-chew committed:
Fix checkstyle issue
